File: Solutions/977-Squares of a Sorted Array/977-Squares of a Sorted Array.cs
//Linq
public class Solution {
    public int[] SortedSquares(int[] A) {
        return A.Select(a => a * a).OrderBy(b => b).ToArray();
    }
}

//Fourth fastest
public class Solution 
{
    public int[] SortedSquares(int[] nums) 
    {
        var n = nums.Length;
        var right = 0;
        
        while(right < n && nums[right] < 0)
        {
            right++;
        }
        
        var left = right - 1;
        
        var ans = new int[n];
        var i = 0;
        
        while(left >= 0 && right < n)
        {
            var leftQ = nums[left]*nums[left];
            var rightQ = nums[right]*nums[right];
            
            if(leftQ < rightQ)
            {
                ans[i] = leftQ;
                left--;
            }
            else
            {
                ans[i] = rightQ;
                right++;
            }
            i++;
        }
        
        while(left >= 0)
        {
            ans[i++] = nums[left]*nums[left];
            left--;
        }
        
        while(right < n)
        {
            ans[i++] = nums[right]*nums[right];
            right++;
        }
        
        return ans;
    }
}

//Third fastest
public class Solution {
    public int[] SortedSquares(int[] nums) {
                    var result = new int[nums.Length];
            var lo = 0;
            var hi = nums.Length - 1;
            var i = nums.Length - 1;
            while (lo <= hi)
            {
                var start = nums[lo] * nums[lo];
                var end = nums[hi] * nums[hi];
                if (start > end)
                {
                    result[i--] = start;
                    lo++;
                }
                else
                {
                    result[i--] = end;
                    hi--;
                }
            }

            return result;
    }
}

//Second fastest
public class Solution {
    public int[] SortedSquares(int[] nums) {
        int[] output = new int[nums.Length];
        int left = 0;
        int right = nums.Length-1;
        int current = nums.Length-1;
        
        while(current >= 0)
        {
            if(Math.Abs(nums[left]) < Math.Abs(nums[right]))
            {
                output[current] = nums[right] * nums[right];
                right--;
            }
            else
            {
                output[current] = nums[left] * nums[left];
                left++;
            }
            current--;
        }
        
        return output;
    }
}

//Fastest
public class Solution {
    public int[] SortedSquares(int[] A) {
        int i = 0;
        int j = A.Length - 1;
        
        while(j-1>=0)
        {
            if(A[j-1] * A[j-1] > A[j] * A[j])
                break;
            j--;
        }
        
        i = j-1;
        
        int[] res = new int[A.Length];
        int index = 0;
        while(i >= 0 && j < A.Length)
        {
            if(A[i] * A[i] < A[j] * A[j])
            {
                res[index] = A[i] * A[i];
                i--;
            }
            else
            {
                res[index] = A[j] * A[j];
                j++;
            }
            index++;
        }
        
        while(i >= 0)
        {
            res[index] = A[i] * A[i];
            i--;
            index++;
        }
        
        while(j < A.Length)
        {
            res[index] = A[j] * A[j];
            j++;
            index++;
        }
        
        return res;
    }
}

File: Solutions/977-Squares of a Sorted Array/977-Squares of a Sorted Array.go
//Third fastest
func sortedSquares(nums []int) []int {
    res := make([]int, len(nums))
    pos := len(nums)-1
    l,r := 0, len(nums)-1
    
    // for i := len(nums)-1; i >= 0; i-- {
    for l <= r {
        ls, rs := nums[l]*nums[l], nums[r]*nums[r]
        if ls < rs {
            res[pos] = rs
            r--
        } else {
            res[pos] = ls
            l++
        }
        pos--
    }
    
    return res

}

// find the first positive number

//Second fastest
import "math"

func sortedSquares(A []int) []int {
    res := make([]int, len(A))
    left, right := 0, len(A)-1
    for end := len(A)-1; end >= 0; end--{
        
        if math.Abs(float64(A[left])) < math.Abs(float64(A[right])) {
            res[end] = A[right]*A[right]
            right -= 1
        } else {
           
            res[end] = A[left]*A[left]
            left += 1
        }
    }
    return res
}

//Fastest
func sortedSquares(A []int) []int {
    i:=0
    j:=0
    t:=0
    tempi:=0
    tempj:=0
    res :=make([]int,len(A))
    
    for i=0;i<len(A) && A[i] < 0 ; {
        i++
    }
    i-=1
    j=i+1
    for i>=0 && j <len(A) {
        tempi=A[i] * A[i]
        tempj=A[j] * A[j]
        if tempi < tempj {
            res[t]=tempi
            i--
        } else {
            res[t]=tempj
            j++
        }
        t++
    }
    
    for i>=0 && t < len(A) {
        res[t] = A[i] * A[i]
        i--
        t++
    }
    
    for j<len(A) {
        res[t]  = A[j] * A[j]
        j++
        t++
    }
    return res
}